### ISSUE TITLE
Dump keychain items?

### ISSUE THREAD

**Issue by a3rosol** [EntySec/SeaShell#13]
**What is your question?**
Is there a way to dump the ios keychains?
If yes, how can I execute it?

If no, could you add this feature?

**Comment by enty8080** [comment]
@a3rosol There is no such feature at the moment, but I will add it.

**Comment by RPwnage** [comment]
I'm pretty sure this will get far more complicated than you might think. Keychain items in their decrypted and readable state are not easily accessible at all.

Might read [this](https://www.theiphonewiki.com/wiki/ICloud_Keychain) and [this](https://jon-gabilondo-angulo-7635.medium.com/understanding-usbmux-and-the-ios-lockdown-service-7f2a1dfd07ae)

**Comment by enty8080** [comment]
@RPwnage If you have any ideas on how to implement such functionality you are free to contribute. I tried to reach you on Twitter, but your DM is closed.

**Comment by RPwnage** [comment]
DM's are open again, feel free to contact me there. I'll look around here in the meantime.

**Comment by enty8080** [comment]
@RPwnage Okay, thank you very much. I saw you were working on EggShell which performs the same thing as SeaShell but for older devices. I tried to make SeaShell more advanced and developer-friendly. However, I have not posted any docs yet. You can also visit this page - [Pwny iOS API](https://github.com/EntySec/Pwny/tree/main/api/ios), it is the implant (payload) that controls the device. You may also take a look at the blog post here - [SeaShell Framework](https://blog.entysec.com/2023-12-31-seashell-ios-malware/).

**Comment by RPwnage** [comment]
Will take a look at this soon 👍 Yes, EggShell was nice to have, but as you‘ve said, it‘s been outdated for some time. I was planning in rewriting EggShell any time soon, but I think that won’t necessary then!

**Comment by enty8080** [comment]
@RPwnage Yeah. Problem with EggShell is not only that it does not support newer versions but also that it was poorly designed from the beginning. I was doing a research and analysed the code of both client side (`esplios` payload) and server side (EggShell). I found multiple ways code could be enhanced. It was better to create a new project from scratch though. What about the payload, it lacks important features. It can't be extended and has no exact negotiation policy.

Pwny, meanwhile, uses [TLV](https://en.wikipedia.org/wiki/Type%E2%80%93length%E2%80%93value) to communicate with server. This way of communication ensures that each packet reaches the destination without any losses. Moreover, Pwny has [`libev`](https://github.com/libev-cn/libev) integrated. It is a powerful C library for asynchronous I/O which allows Pwny to not read socket in loop, but wait for the event occur and read only then. This enhanced efficiency and resource utilisation. I will continue working on documenting Pwny and since there are lots of C code, I hope to make it understandable.

I guess one of the main advantages of SeaShell is that its payload can work in non-jailbroken environment. All you need to have is CoreTrust bypass.

**Comment by RPwnage** [comment]
I might already have a solution handy to decrypting the keychain .db now that I think about it again. If you want to further discuss implementing it, you can always message me on twitter. My DMs are open again.

**Comment by enty8080** [comment]
@rpwnage I DMed you in Twitter

**Comment by joyas11** [comment]
> @rpwnage I DMed you in Twitter

any updates? can it be implemented

**Comment by enty8080** [comment]
@joyas11 Probably yes, however I am still waiting for update from @rpwnage on this issue

**Comment by jacobprezant** [comment]
Any updates on this? I would be interested in working on it

**Comment by enty8080** [comment]
@jacobprezant Thats an interesting thing to work on, I welcome any help on this issue :)

**Comment by jacobprezant** [comment]
Did you make any progress with @rpwnage? @enty8080

**Comment by enty8080** [comment]
No, he just disappeared
